RecoHGCal/TICL/plugins/PatternRecognitionbyKF.cc

```diff
@@ -0,0 +1,481 @@
+                                              const TILES &tiles, PropagationDirection direction, 
+                                              bool &isSilicon, 
+                                              TempTrajectory traj){
+
+  std::vector<TempTrajectory> ret;
+
+  const Propagator &prop = (direction == alongMomentum ? *propagator_ : *propagatorOppo_);
+  TrajectoryStateOnSurface tsos = prop.propagate(start, disk->surface());
+  if (!tsos.isValid()) return ret;
+  float r = sqrt(pow(tsos.globalPosition().x(),2)+pow(tsos.globalPosition().y(),2));
+  if (((disk->rmin() > r) && (!isSilicon)) || (((r > disk->rmax()) && (isSilicon)))) {
+    disk = switchDisk(disk, disks, isSilicon);
+    isSilicon = !isSilicon;
+    tsos = prop.propagate(start, disk->surface());
+  }
+
+  // Collect hits with estimate
+  int depth = disk->layer()+1;
+  auto meas = measurements(tsos, *estimator_, tiles, depth);
+  std::sort(meas.begin(), meas.end(),TrajMeasLessEstim());
+
+  for (const TrajectoryMeasurement &tm : meas){
+    TrajectoryStateOnSurface updated = updator_->update(tm.forwardPredictedState(),*tm.recHit());
+    ret.push_back(traj.foundHits() ? traj: TempTrajectory(traj.direction(),0));
+    ret.back().push(TrajectoryMeasurement(tm.forwardPredictedState(),
+                                          updated,
+                                          tm.recHit(),
+                                          tm.estimate()),
+                    tm.estimate());
+  }
+
+  if (!meas.empty()) return ret;
+  auto missing = TrackingRecHit::missing;
+  ret.push_back(traj.foundHits()? traj : TempTrajectory(traj.direction(),0));
+  ret.back().push(TrajectoryMeasurement(tsos, std::make_shared<InvalidTrackingRecHit>(*disk,missing)));
+
+  return ret;
+}
+
+template<typename TILES>
+void PatternRecognitionbyKF<TILES>::makeDisks(int subdet, int disks, const CaloGeometry* geom_) {
+
+    const CaloSubdetectorGeometry *subGeom = geom_->getSubdetectorGeometry(DetId::Detector(subdet), ForwardSubdetector::ForwardEmpty);
+    auto geomEE = static_cast<const HGCalGeometry*>(subGeom);
+    const HGCalDDDConstants* ddd = &(geomEE->topology().dddConstants());
+
+    std::vector<float>  rmax(disks, 0), rmin(disks, 9e9);
+    std::vector<double> zsumPos(disks), zsumNeg(disks);
+    std::vector<int> countPos(disks), countNeg(disks);
+    const std::vector<DetId> & ids = subGeom->getValidDetIds();
+  
+    for (auto & i : ids) {
+        calculateLocalError(i,geomEE);
+
+        const GlobalPoint & pos = rhtools_.getPosition(i); 
+        int layer = rhtools_.getLayer(i)-1;
+        float z = pos.z();
+        float rho = pos.perp();
+        int side = z > 0 ? +1 : -1;
+
+        (side > 0 ? zsumPos : zsumNeg)[layer] += z;
+        (side > 0 ? countPos : countNeg)[layer]++;
+        if (rho > rmax[layer]) rmax[layer] = rho;
+        if (rho < rmin[layer]) rmin[layer] = rho;
+    }
+
+  int layer = ddd->getLayerOffset(); // FIXME: Can be made less stupid
```

Comment:
stupid question: why this and not `firstLayer()`: https://github.com/cms-sw/cmssw/blob/master/Geometry/HGCalCommonData/interface/HGCalDDDConstants.h#LL70C23-L70C23
?